packages/react-stand-in/src/createClassProxy.js

```diff
@@ -1,16 +1,32 @@
-import { Component } from 'react'
+import { Component, createElement } from 'react'
 import transferStaticProps from './transferStaticProps'
 import { GENERATION, PROXY_KEY, UNWRAP_PROXY } from './constants'
 import {
   getDisplayName,
   isReactClass,
+  isReactElement,
   identity,
   safeDefineProperty,
 } from './utils'
 import { inject, checkLifeCycleMethods, mergeComponents } from './inject'
 
 const proxies = new WeakMap()
 
+function Dereference(props) {
+  // copy over pros
+  const classInstance = props.standin_classInstance
+  classInstance.props = props
+  return classInstance
+}
+Dereference.DO_NOT_HOT_RELOAD = true
+
+function wrapWithStateless(classInstance, props) {
```

Comment:
I had to read this a few times to figure out what this was doing. It's not exactly wrapping this with a stateless functional component – it's more like `rewrapAsIndeterminate`.

Reply:
It just creates `a real` functional component, which will be Interminate until the execution, and can carry desired payload.

Reply:
This isn't a standard functional component, though. In this case, the return value is not a React element – it's a React component instance.

As such, it goes through that "indeterminate" code path, and React realizes that this is actually a class component.

Reply:
Yet again - just remove RHL from you mind.
1. Reproduce the problem.
 You will found that you can return renderable-component only from Stateless functional, not from class-based. (and there is no third way).
2. Let RHL to mimic same behaviour.
 RHL removes functional components, making all class-based - thats is the problem. We have to do it, cos this is the only(*) way to be able hot-swap function implementation.

This means - if you found some odd behavior, you could reproduce only using functional component - use functional component to reproduce the original behavior. 
This also means - we should wrap functional components with functions. It will be the best way to solve any further issues.

Reply:
+1 on that last point, probably the only way that will work in most cases (save for a function storing state in its scope or at module level)

Reply:
Just spiked possible solution with keeping functions as functions - it is much more clear, but results 5% tests to fail.
I'll hope to work it out till the end of the day.

Reply:
@theKashey 

Take a look at this CodeSandbox: https://codesandbox.io/s/6wz0or0rqk

This is not exactly a stateless functional component. It gets resolved as a normal class-based component.

In the first set of commands, if you look at the console output, you'll see `FooConstructor` is only called once, while `Foo.render` is called 3 times. If you look at https://6wz0or0rqk.codesandbox.io/ with React Dev Tools, you'll see that a single `FooConstructor` component gets mounted, and it is a full-fledged class component with state and everything.

By contrast, if you uncomment out the render calls with the proxy, you'll see that:
- The component tree is different
- `FooConstructor` gets called 3 times instead of just 1 time

So it's not the same. But in both these cases, no stateless functional components are getting rendered.

Reply:
i.e.

![image](https://user-images.githubusercontent.com/3112159/34809340-5e328822-f662-11e7-83d9-5752cfd90dac.png)

For

```js
import React from 'react';
import ReactDOM from 'react-dom';

class Foo extends React.Component {
  constructor(props, context) {
    super(props, context);

    this.state = {
      foo: 'bar',
    };
  }

  render() {
    console.log('foo');
    return <div>foo</div>;
  }
}

function FooConstructor(props, context) {
  console.log('foo constructor');
  return new Foo(props, context);
}

ReactDOM.render(<FooConstructor />, document.getElementById('root'));
ReactDOM.render(<FooConstructor />, document.getElementById('root'));
ReactDOM.render(<FooConstructor />, document.getElementById('root'));
```

There are no stateless functional components involved here.

Reply:
Unfortunately, we have to proceed with this approach.
I've spiked the other way to wrapping the stuff, keeping functional components functional, and works with this case without any magic, but....
But stateless component does not leave `instance` behind, preventing _hot-reconciler_ to properly work. 

To fix this one have to rework `getReactStack` to properly handle any possible React.Element type.

Reply:
See this sandbox: https://codesandbox.io/s/jl3267871y

Reply:
During the hot replace we start process to merge the old and the new react tree, to update components which are not exposed as variables or just hidden behind decorators. 
During this process we can’t handle function components, except this case, as long React stores only return variable in the tree. 
Your code example is eligible, but hotReplacement can’t pass though the real functional component. 

For now - everything except this case have to be wrapped with class.